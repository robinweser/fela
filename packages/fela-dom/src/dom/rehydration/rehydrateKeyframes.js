import { KEYFRAME_TYPE } from 'fela-utils'

const RE = /@(-webkit-|-moz-)?keyframes ([a-z_][a-z0-9-_]*)(\{.*?(?=}})}})/gi

export default function rehydrateKeyframes(css, cache = {}) {
  const matches = css.matchAll(RE)

  for (const match of matches) {
    const [keyframe, , animationName, reference] = match

    if (!cache[reference]) {
      cache[reference] = {
        type: KEYFRAME_TYPE,
        keyframe,
        name: animationName,
      }
    } else {
      cache[reference].keyframe += keyframe
    }
  }

  return cache
}
