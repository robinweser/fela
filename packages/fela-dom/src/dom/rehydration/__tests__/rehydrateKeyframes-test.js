import rehydrateKeyframes from '../rehydrateKeyframes'

describe('Rehydrating keyframes', () => {
  it('should match all keyframes', () => {
    expect(
      rehydrateKeyframes(
        '@keyframes frameName1{0% {opacity:0;}}; @keyframes frameName2{100%{opacity: 1;}}'
      )
    ).toMatchSnapshot()
  })
})
