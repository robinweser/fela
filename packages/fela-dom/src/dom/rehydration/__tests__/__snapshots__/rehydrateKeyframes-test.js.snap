// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rehydrating keyframes should match all keyframes 1`] = `
{
  "{0% {opacity:0;}}": {
    "keyframe": "@keyframes frameName1{0% {opacity:0;}}",
    "name": "frameName1",
    "type": "KEYFRAME",
  },
  "{100%{opacity: 1;}}": {
    "keyframe": "@keyframes frameName2{100%{opacity: 1;}}",
    "name": "frameName2",
    "type": "KEYFRAME",
  },
}
`;
